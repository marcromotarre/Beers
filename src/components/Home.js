import React, { useState, useEffect } from 'react';
import { useSelector } from 'react-redux';
import styled from 'styled-components';
import { useDispatch } from 'react-redux';

import * as types from '../redux/constants';
import Pagination from './Pagination';
import BeersList from './Beer/BeersList';
import BrewedBetween from './Calendar/BrewedBetween';
import {
  getMinBrewedDate,
  getMaxBrewedDate,
  getBeersByBrewedFromTo,
} from '../utils/filters';

const HomeContainer = styled.div`
  display: grid;
  grid-template-columns: 100%;
  grid-template-rows: 250px auto 100px;
  height: 100vh;

  > * {
    justify-self: center;
    align-self: center;
  }
`;

const getBeersByPage = (beers, page, numberElementsPerPage) => {
  return beers.filter(
    (beer, index) =>
      index >= (page - 1) * numberElementsPerPage &&
      index < page * numberElementsPerPage
  );
};

const Home = () => {
  const dispatch = useDispatch();
  const { beers } = useSelector((state) => state.beersReducer);
  const { brewedFrom, brewedTo } = useSelector((state) => state.filtersReducer);

  const numberElementsPerPage = 3;
  const defaultPageNumber = 1;
  const [page, setPage] = useState(defaultPageNumber);

  useEffect(() => {
    dispatch({
      type: types.setBrewedFromTo,
      brewed: {
        from: getMinBrewedDate(beers),
        to: getMaxBrewedDate(beers),
      },
    });
  }, []);

  const filteredBeers = getBeersByBrewedFromTo({ beers, brewedFrom, brewedTo });
  const setPageNumber = (clickedPage) => {
    setPage(clickedPage);
  };

  return (
    <>
      <HomeContainer data-testid="home-container">
        <BrewedBetween />
        <BeersList
          beers={getBeersByPage(filteredBeers, page, numberElementsPerPage)}
        ></BeersList>
        {filteredBeers.length > 0 && (
          <Pagination
            numberTotalElements={filteredBeers.length}
            numberElementsPerPage={numberElementsPerPage}
            defaultPageNumber={defaultPageNumber}
            onClickPage={setPageNumber}
          />
        )}
      </HomeContainer>
    </>
  );
};

export default Home;
