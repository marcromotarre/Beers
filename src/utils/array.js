export const generateAutoIncrementalArrayFromMinToMax = (min, max) => {
  return new Array(max - min + 1).fill().map((number, index) => min + index);
};
