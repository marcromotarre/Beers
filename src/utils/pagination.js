export const getNumPages = ({ totalElements, elementsPerPage }) => {
  return Math.ceil(totalElements / elementsPerPage);
};

export const generatePageNumbers = ({
  actualPage,
  numPages,
  maxNumShowedPages,
}) => {
  const pagesPerSide = Math.floor(maxNumShowedPages / 2);
  const arrayPages = new Array(numPages)
    .fill()
    .map((number, index) => index + 1);

  if (numPages <= maxNumShowedPages) return { pages: arrayPages };
  if (actualPage - pagesPerSide <= 1) {
    return {
      last: numPages,
      pages: arrayPages.filter((page) => page < maxNumShowedPages - 1),
    };
  }
  if (actualPage + pagesPerSide >= numPages) {
    return {
      first: 1,
      pages: arrayPages.filter(
        (page) => page > numPages - (maxNumShowedPages - 2)
      ),
    };
  }

  return {
    first: 1,
    pages: arrayPages.filter(
      (page) =>
        page < actualPage + (pagesPerSide - 1) &&
        page > actualPage - (pagesPerSide - 1)
    ),
    last: numPages,
  };
};
