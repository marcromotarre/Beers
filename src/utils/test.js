import React from 'react';
import { render } from '@testing-library/react';

import {
  createHistory,
  createMemorySource,
  LocationProvider,
} from '@reach/router';

import { createStore } from 'redux';
import reducer from './../redux/reducers/index';
import { Provider } from 'react-redux';

export function renderRouteRedux(
  ui,
  {
    route = '/',
    history = createHistory(createMemorySource(route)),
    initialState,
    store = createStore(reducer, initialState),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return (
      <Provider store={store}>
        <LocationProvider history={history}>{children}</LocationProvider>
      </Provider>
    );
  }
  return render(ui, { wrapper: Wrapper, ...renderOptions });
}

export function renderRedux(
  ui,
  {
    initialState,
    store = createStore(reducer, initialState),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>;
  }
  return render(ui, { wrapper: Wrapper, ...renderOptions });
}

export function renderReduxWithRouter(
  ui,
  {
    route = '/',
    history = createHistory(createMemorySource(route)),
    initialState,
    store = createStore(reducer, initialState),
    ...renderOptions
  } = {}
) {
  return {
    ...render(
      <Provider store={store}>
        <LocationProvider history={history}>{ui}</LocationProvider>
      </Provider>
    ),
    history,
  };
}
