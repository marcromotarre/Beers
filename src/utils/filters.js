import { formatBrewed } from './text';
import { generateAutoIncrementalArrayFromMinToMax } from './array';

export const getBeersByBrewedFromTo = ({
  beers = [],
  brewedFrom,
  brewedTo,
}) => {
  return beers.filter((beer) => {
    const brewed = formatBrewed(beer.first_brewed);
    return (
      (brewed.year > brewedFrom.year && brewed.year < brewedTo.year) ||
      (brewed.year === brewedFrom.year &&
        brewed.month >= brewedFrom.month &&
        brewed.year < brewedTo.year) ||
      (brewed.year > brewedFrom.year &&
        brewed.year === brewedTo.year &&
        brewed.month <= brewedTo.month) ||
      (brewed.year === brewedFrom.year &&
        brewed.year === brewedTo.year &&
        brewed.month >= brewedFrom.month &&
        brewed.month <= brewedTo.month)
    );
  });
};

export const getMinBrewedDate = (beers = []) => {
  return beers
    .map((beer) => formatBrewed(beer.first_brewed))
    .sort(function (a, b) {
      return a.year <= b.year && a.month <= b.month ? -1 : 1;
    })[0];
};

export const getMaxBrewedDate = (beers) => {
  return beers
    .map((beer) => formatBrewed(beer.first_brewed))
    .sort(function (a, b) {
      return a.year <= b.year && a.month <= b.month ? 1 : -1;
    })[0];
};

export const getYearsFromMinToMaxBrewed = (beers) => {
  return generateAutoIncrementalArrayFromMinToMax(
    getMinBrewedDate(beers).year,
    getMaxBrewedDate(beers).year
  );
};

export const getMonths = () => {
  return generateAutoIncrementalArrayFromMinToMax(1, 12);
};
