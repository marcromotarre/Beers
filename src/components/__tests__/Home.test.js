import React from 'react';
import renderer from 'react-test-renderer';
import { render } from '@testing-library/react';
import Home from '../Home';
import { renderRedux } from '../../utils/test';

import beers from '../../fixtures/beers';

/** this component should render:
 *  -  a list of beers:
 *          or
 *  - label in casse the list is empty
 *
 */

const initialState = {
  beersReducer: { pending: false, error: false, beers },
};

describe('Home Component Snapshots', () => {
  it('Home Snapshots', () => {
    expect(
      renderRedux(<Home />, {
        initialState,
      }).baseElement
    ).toMatchSnapshot();
  });
});

describe('Home Component tests', () => {
  it('should render calendar', () => {
    const { getByTestId } = renderRedux(<Home />, {
      initialState,
    });

    expect(getByTestId('calendar-month-year-view-2007-9')).toBeTruthy();
    expect(getByTestId('calendar-month-year-view-2015-11')).toBeTruthy();
    expect(getByTestId('beers-list-container').children.length).toBeTruthy();
  });

  it('should render beers list', () => {
    const { getByTestId } = renderRedux(<Home />, {
      initialState,
    });
    expect(getByTestId('beers-list-container').children.length).toBe(3);
  });

  it('should render pagination list', () => {
    const { getByTestId } = renderRedux(<Home />, {
      initialState,
    });
    const pagination = getByTestId('pagination-container');
    expect(pagination.children.length).toBe(4);
  });
});
