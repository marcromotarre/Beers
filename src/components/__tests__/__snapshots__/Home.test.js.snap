// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Component Snapshots Home Snapshots 1`] = `
<body>
  <div>
    <div
      class="sc-fzoyAV jHhSSh"
      data-testid="home-container"
    >
      <section
        class="sc-fzokOt eyyxQb"
        data-testid="brewed-between-container"
      >
        <section
          class="sc-fzqBZW esvDpU"
        >
          <button
            data-testid="calendar-month-year-view-2007-9"
          >
            <section
              class="sc-fzozJi hiluBJ"
            >
              <div
                class="sc-fzpans hSeYyq"
                height="68.4931506849315px"
                width="100px"
              >
                <div
                  class="sc-fznyAO WMTwo"
                  data-testid="calendar-month-year-view-container"
                >
                  <label
                    class="sc-fznKkj hFPand"
                  >
                    SEPTEMBER
                  </label>
                  <label
                    class="sc-fznZeY idMIqi"
                  >
                    2007
                  </label>
                </div>
              </div>
              <svg
                class="sc-fzoLsD fqEnxe"
                width="100"
              />
            </section>
          </button>
          <div
            class="sc-fzqNJr gdryOv"
          >
            <label>
              brewed
            </label>
            <svg
              width="28"
            />
            <label>
              between
            </label>
          </div>
          <button
            data-testid="calendar-month-year-view-2015-11"
          >
            <section
              class="sc-fzozJi hiluBJ"
            >
              <div
                class="sc-fzpans hSeYyq"
                height="68.4931506849315px"
                width="100px"
              >
                <div
                  class="sc-fznyAO WMTwo"
                  data-testid="calendar-month-year-view-container"
                >
                  <label
                    class="sc-fznKkj hFPand"
                  >
                    NOVEMBER
                  </label>
                  <label
                    class="sc-fznZeY idMIqi"
                  >
                    2015
                  </label>
                </div>
              </div>
              <svg
                class="sc-fzoLsD fqEnxe"
                width="100"
              />
            </section>
          </button>
        </section>
      </section>
      <section
        class="sc-AxmLO hOHgZI"
        data-testid="beers-list-container"
      >
        <section
          class="sc-AxiKw gRdjzo"
          data-testid="beer-container"
        >
          <div
            class="sc-AxhCb cxbxdi"
          >
            <div
              class="sc-AxhUy iqRDDD"
            >
              <h1>
                Buzz
              </h1>
              <h2>
                A Real Bitter Experience.
              </h2>
              <p>
                A light, crisp and bitter IPA brewed with English and American hops. A small batch brewed only once.
              </p>
            </div>
            <div
              class="sc-AxgMl coxPBi"
            >
              <label>
                brewed:
              </label>
              <p>
                09/2007
              </p>
            </div>
          </div>
          <div
            class="sc-AxheI foprKJ"
          >
            <img
              src="https://images.punkapi.com/v2/keg.png"
            />
          </div>
        </section>
        <section
          class="sc-AxiKw gRdjzo"
          data-testid="beer-container"
        >
          <div
            class="sc-AxhCb cxbxdi"
          >
            <div
              class="sc-AxhUy iqRDDD"
            >
              <h1>
                Trashy Blonde
              </h1>
              <h2>
                You Know You Shouldn't
              </h2>
              <p>
                A titillating, neurotic, peroxide punk of a Pale Ale. Combining attitude, style, substance, and a little bit of low self esteem for good measure; what would your mother say? The seductive lure of the sassy passion fruit hop proves too much to resist. All that is even before we get onto the fact that there are no additives, preservatives, pasteurization or strings attached. All wrapped up with the customary BrewDog bite and imaginative twist.
              </p>
            </div>
            <div
              class="sc-AxgMl coxPBi"
            >
              <label>
                brewed:
              </label>
              <p>
                04/2008
              </p>
            </div>
          </div>
          <div
            class="sc-AxheI foprKJ"
          >
            <img
              src="https://images.punkapi.com/v2/2.png"
            />
          </div>
        </section>
        <section
          class="sc-AxiKw gRdjzo"
          data-testid="beer-container"
        >
          <div
            class="sc-AxhCb cxbxdi"
          >
            <div
              class="sc-AxhUy iqRDDD"
            >
              <h1>
                Berliner Weisse With Yuzu - B-Sides
              </h1>
              <h2>
                Japanese Citrus Berliner Weisse.
              </h2>
              <p>
                Japanese citrus fruit intensifies the sour nature of this German classic.
              </p>
            </div>
            <div
              class="sc-AxgMl coxPBi"
            >
              <label>
                brewed:
              </label>
              <p>
                11/2015
              </p>
            </div>
          </div>
          <div
            class="sc-AxheI foprKJ"
          >
            <img
              src="https://images.punkapi.com/v2/keg.png"
            />
          </div>
        </section>
      </section>
      <section
        class="sc-AxjAm kMuvoV"
        data-testid="pagination-container"
      >
        <button
          disabled=""
        />
        <button
          id="1"
        >
          <svg
            class="sc-AxirZ iuEUCp"
            height="60"
          />
          <label
            class="selected"
          >
            1
          </label>
        </button>
        <button
          id="2"
        >
          <label
            class=""
          >
            2
          </label>
        </button>
        <button>
          <svg
            width="16.43"
          />
        </button>
      </section>
    </div>
  </div>
</body>
`;
